library/init/meta/expr.lean

```diff
@@ -247,13 +247,10 @@ is_bin_arith_app e `le
 meta def is_ge (e : expr) : option (expr × expr) :=
 is_bin_arith_app e `ge
 
-meta def is_heq (e : expr) : option (expr × expr × expr × expr) :=
-if is_napp_of e `heq 4
-then some (app_arg (app_fn (app_fn (app_fn e))),
-           app_arg (app_fn (app_fn e)),
-           app_arg (app_fn e),
-           app_arg e)
-else none
+set_option eqn_compiler.max_steps 512
+meta def is_heq : expr → option (expr × expr × expr × expr)
+| ``((%%a: %%α) == (%%b : %%β)) := some (α, a, β, b)
+| _                             := none
```

Comment:
The new definitions look much nicer, but what is the performance impact?
It it is too high, and we can try to optimize the pattern matching compilation code.